sys/newlib/syscalls.c

```diff
@@ -85,15 +85,14 @@ void _exit(int n)
  *
  * @return [description]
  */
-caddr_t _sbrk_r(struct _reent *r, ptrdiff_t incr)
+void *_sbrk_r(struct _reent *r, ptrdiff_t incr)
 {
     unsigned int state = disableIRQ();
-    caddr_t res = heap_top;
+    void *res = heap_top;
 
-    if (((incr > 0) && ((heap_top + incr > &_eheap) || (heap_top + incr < res))) ||
-        ((incr < 0) && ((heap_top + incr < &_sheap) || (heap_top + incr > res)))) {
```

Comment:
It was doing two different comparisons depending on the sign of `incr`. The comparison against `res` was probably intended to catch overflows.
In the new code, overflows will always be caught, except for the case where `_eheap - _sheap > (1 << sizeof(ptrdiff_t)) / 2` and `incr` is large enough to overflow and still make the heap pointer end up inside the valid heap area.